fix opencv/opencv#20486 switching opencl contexts

diablodale opened the pull request:
Series of fixes opencv/opencv#20486
NOT YET READY FOR MERGE.
And it is currenty based on 4.5.3 (not tested or based on 3.x)

### Pull Request Readiness Checklist

- [X] I agree to contribute to the project under Apache 2 License.
- [X] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [ ] The PR is proposed to proper branch
- [X] There is reference to original bug report and related work
- [X] The feature is well documented and sample code can be built with the project CMake

alalek commented:
Hi, current patch has huge memory leak. Please test it locally before submitting (it breaks our CI)

diablodale commented:
Hi. I have been testing here on three platforms (win10, ubuntu16 without svm, and ubuntu16 with svm). I regularily run core, imgproc, and stitching test suites for unit testing. I have not reach any performance tests yet.

I'll look again but haven't had the issues I've seen in the CI recently. Some of the CI problems I saw were during compiling, and before programs ran in diectory/file copying. How can SDK code I write affect the CI's ability to copy a file? 🤷‍♀️

alalek commented:
@diablodale Hi! What is the status of this PR? It would be great if you can submit ready-to-merge small parts.

diablodale commented:
Hi. The PR is alive and working well with my app and use cases. 😉

🙏 I request you, me, and someone else on your core maintainer team have an online chat together. I believe a discussion on this work is nuanced and will benefit with real-time conversation. My goal of a chat is to have progress/answers on all the issues below -- naturally documenting our outcome in issues/PR/code.

* Locally, I am 5 commits ahead so the code here in this PR is not complete. I haven't pushed more commits because there are several bugs/issues that block or strongly hinders this PR's work a success.
* I finished almost all the functional changes I intended in this "first refactoring" but I'm blocked as I describe further below
* I believe the PR also needs a optimization pass as it relates to the read/readwrite/obsolete behavior. I suspect there is a design flaw or bug in this area. I paused investigation/work on that due to the issues below.

### Grouped Issues, ranked from blocking -> least 

1. issue #20615 is blocking. Until this is fixed, all test results of the DNN module are unreliable. They can report false positives and false negatives. DNN is a heavy user of OpenCL. Therefore, part of the validation of this PR's work is to pass the DNN tests -- which are currently themselves unreliable.
2. The group of RFC issues #20583, #20582, #20577, #20565. I have already incorporated fixes into this PR for some of these because I needed working/reliable code for my app. If the team choses to take a different approach on those RFC, then the PR will need to be reworked.
3. Issue #20542 has some work within itself but is first dependent on some RFC issue answers.
4. External bug https://developer.nvidia.com/nvidia_bug/3360381 blocks all NVidia OpenCL SVM use cases. This PR's refactoring fixed SVM so it can now be used across modules and GPU manufacturers. But, this NVidia driver bug causes widespread failure of OpenCL when SVM is used. I have a code in this PR to disable Nvidia+OpenCL+SVM. NVidia identified and fixed the issue but has not yet released the fix in a public driver. I prefer to have the fixed driver for comprehensive testing of both SVM and of the code that disables Nvidia+SVM bad driver versions. This isn't blocking, but since other issues are blocking, this is not the critical path.
4. External bug https://github.com/microsoft/vscode-cpptools/issues/5548#issuecomment-926895244 is not blocking but makes my life easier.
